File: CMPS101/pa2/List.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "List.h"

// ** Structs **

typedef struct NodeObj 
{
    int data;
    struct NodeObj* next;
    struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj
{
    Node front, back, cursor;
    int index, length;
} ListObj;


// ** Constructors-destructors **

// Creates new Node.
Node newNode(int data)
{
    Node n = malloc(sizeof(NodeObj));
    n->data = data;
    n->next = NULL;
    n->prev = NULL;
    return n;
}

// Creates new Node with prev and next parameters
Node newNodePN(int data, Node prev, Node next)
{
    Node n = malloc(sizeof(NodeObj));
    n->data = data;
    n->next = next;
    n->prev = prev;
    return n;
}

// Frees memory of pN, and sets *pN to NULL.
void freeNode(Node* pN)
{
    if (pN != NULL && *pN != NULL)
    {
        free(*pN);
        *pN = NULL;
    }
}

// Recursively frees all nodes starting at pN.
void freeAllNodes(Node* pN)
{
    if (pN != NULL && *pN != NULL)
    {
        if ((*pN)->next != NULL)
            freeAllNodes(&((*pN)->next));
        free(*pN);
        *pN = NULL;
    }
}

// List constructor
List newList(void)
{
    List L = malloc(sizeof(ListObj));
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->length = 0;
    L->index = -1;
    return L;
}

// Removes all memory associated with the List.
// Also needs to delete all Nodes within the list, I think.
void freeList(List* pL)
{
    if (pL != NULL && *pL != NULL)
    {
        // Remove all node memory from list.
        freeAllNodes(&(*pL)->front);

        free(*pL);
        *pL = NULL;
    }
}

// ** Access functions **

// Return length of the list
int length(List L)
{
    if (L == NULL) {
        printf("List Error: length(): L is NULL.\n");
        exit(1);
    }
    return L->length;
}

// Return index of cursor. If not defined, return -1.
int index(List L)
{
    if (L == NULL) {
        printf("List Error: index(): L is NULL.\n");
        exit(1);
    }
    if (L->cursor == NULL)
        return -1;
    else
        return L->index;
}

// Pre: length > 0
// Return data at front of list
int front(List L)
{
    if (L == NULL) {
        printf("List Error: front(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: front(): length <= 0.\n");
        exit(1);
    }
    return L->front->data;
}

// Pre: length > 0
// Return data at back of list.
int back(List L)
{
    if (L == NULL) {
        printf("List Error: back(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: back(): length <= 0.\n");
        exit(1);
    }
    return L->back->data;
}

// Pre: length > 0
// Return data at cursor position
int get(List L)
{
    if (L == NULL) {
        printf("List Error: get(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0)
    {
        printf("List Error: get(): length <= 0.\n");
        exit(1);
    }
    return L->cursor->data;
}

// Returns 1 if lists are equal, 0 if they are not.
// Checks if all elements in the lists are the same in the same positions
int equals(List A, List B)
{
    if (A == NULL || B == NULL) {
        printf("List Error: equals(): A or B is NULL.\n");
        exit(1);
    }

    Node cur = A->front;
    Node LCur = B->front;

    while (cur != NULL)
    {
        // if B is null before A, then it's necessarily not equal
        if (LCur == NULL || LCur->data != cur->data)
            return 0;
        cur = cur->next;
        LCur = LCur->next;
    }
    return 1;
}

// ** Manipulation procedures **

// Reset list to empty
void clear(List L)
{
    if (L == NULL) {
        printf("List Error: clear(): L is NULL.\n");
        exit(1);
    }
    freeNode(&(L->cursor));
    // Free all nodes from start to end.
    freeAllNodes(&(L->front));
    L->index = -1;
    L->length = 0;
}

// If list not empty, put cursor at front.
// Else, do nothing
void moveFront(List L)
{
    if (L == NULL) {
        printf("List Error: moveFront(): L is NULL.\n");
        exit(1);
    }
    if (length(L) > 0)
    {
        L->cursor = L->front;
        L->index = 0;
    }
}

// If list not empty, put cursor at back
// Else, do nothing
void moveBack(List L)
{
    if (L == NULL) {
        printf("List Error: moveBack(): L is NULL.\n");
        exit(1);
    }
    if (length(L) > 0)
    {
        L->cursor = L->back;
        L->index = length(L) - 1;
    }
}

// If cursor defined, if not at beginning of list, move cursor back. Else, set cursor to undefined. Else, do nothing
void movePrev(List L)
{
    if (L == NULL) {
        printf("List Error: movePrev(): L is NULL.\n");
        exit(1);
    }
    if (L->cursor != NULL)
    {
        if (index(L) > 0)
        {
            L->cursor = L->cursor->prev;
            L->index--;
        }
        else
        {
            L->cursor = NULL;
        }
    }
}

// If cursor defined, if not at end of list, move cursor forward. Else, set cursor to undefined. Else, do nothing
void moveNext(List L)
{
    if (L == NULL) {
        printf("List Error: moveNext(): L is NULL.\n");
        exit(1);
    }
    if (L->cursor != NULL)
    {
        if (index(L) != length(L) - 1)
        {
            L->cursor = L->cursor->next;
            L->index++;
        }
        else
        {
            L->cursor = NULL;
        }
    }
}

// Insert new element.
//If list is non-empty, put before front element
void prepend(List L, int data)
{
    if (L == NULL) {
        printf("List Error: prepend(): L is NULL.\n");
        exit(1);
    }
    Node n = newNodePN(data, NULL, L->front);
    if (length(L) == 0)
        L->back = n;
    if (L->front != NULL)
        L->front->prev = n;
    L->front = n;
    L->length++;
}

// Insert new element.
// If list is non-empty, put after back element
void append(List L, int data)
{
    if (L == NULL) {
        printf("List Error: append(): L is NULL.\n");
        exit(1);
    }
    Node n = newNodePN(data, L->back, NULL);
    if (length(L) == 0)
        L->front = n;
    if (L->back != NULL)
        L->back->next = n;
    L->back = n;
    L->length++;
    //printf("Appended to L w/ %d\n", data);
}

// Insert new element before cursor
// Pre: length>0, index>=0
void insertBefore(List L, int data)
{
    if (L == NULL) {
        printf("List Error: insertBefore(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0) {
        printf("List Error: insertBefore(): length <= 0\n");
        exit(1);
    }
    if (index(L) < 0) {
        printf("List Error: insertBefore(): index < 0\n");
        exit(1);
    }

    Node n = newNodePN(data, L->cursor->prev, L->cursor);
    if (L->cursor->prev != NULL)
        L->cursor->prev->next = n;
    else // There was nothing before the cursor, cursor was at front, and we need to update front
        L->front = n;
    L->cursor->prev = n;
    L->length++;
}

// Insert new element after cursor
// Pre: length>0, index>=0
void insertAfter(List L, int data)
{
    if (L == NULL) {
        printf("List Error: insertAfter(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0) {
        printf("List Error: insertAfter(): length <= 0\n");
        exit(1);
    }
    if (index(L) < 0) {
        printf("List Error: insertAfter(): index < 0\n");
        exit(1);
    }

    Node n = newNodePN(data, L->cursor, L->cursor->next);
    if (L->cursor->next != NULL)
        L->cursor->next->prev = n;
    else // If there was nothing after the cursor, cursor was at back, and we need to update back
        L->back = n;
    L->cursor->next = n;
    L->length++;
}

// Delete front element
// Pre: length>0
void deleteFront(List L)
{
    if (L == NULL) {
        printf("List Error: deleteFront(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0) {
        printf("List Error: deleteFront(): length <= 0\n");
        exit(1);
    }
    // We can't do the same as in java and dereference things, unfortunately
    L->front = L->front->next;
    freeNode(&(L->front->prev));
    L->front->prev = NULL;
    if (index(L) == 0) // If cursor was at front, repoint it to the current front
    {
        L->cursor = L->front;
    }
    L->length--;
}

// Delete back element
// Pre: length>0
void deleteBack(List L)
{
    if (L == NULL) {
        printf("List Error: deleteBack(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0) {
        printf("List Error: deleteBack(): length <= 0\n");
        exit(1);
    }
    // Clear back node
    L->back = L->back->prev;
    freeNode(&(L->back->next));
    if (index(L) == length(L) - 1) // If cursor was at back, repoint it to new back
    {
        L->cursor = L->back;
    }
    L->length--;
    L->index--; // Decrease index by 1 because we lost an element
}

// Delete cursor element, making cursor undefined
// Pre: length>0, index>=0
void delete(List L)
{
    if (L == NULL) {
        printf("List Error: delete(): L is NULL.\n");
        exit(1);
    }
    if (length(L) <= 0) {
        printf("List Error: delete(): length <= 0\n");
        exit(1);
    }
    if (index(L) < 0) {
        printf("List Error: delete(): index < 0\n");
        exit(1);
    }

    L->cursor->prev->next = L->cursor->next;
    L->cursor->next->prev = L->cursor->prev;
    L->length--;
    L->cursor = NULL;
    if (index(L) > 0)
        L->index = -1; // Reset index to -1 since cursor is undefined
}

// ** Other functions **

void printList(FILE* out, List L)
{
    if (L == NULL) {
        printf("List Error: printList(): L is NULL.\n");
        exit(1);
    }
    if (out == NULL) {
        printf("List Error: printlist(): out is NULL.\n");
        exit(1);
    }

    Node cur = L->front;
    while (cur != NULL)
    {
        fprintf(out, "%d ", cur->data);
        cur = cur->next;
    }
}

List copyList(List L)
{
    List nL = newList();
    Node cur = L->front;

    while (cur != NULL)
    {
        append(nL, cur->data);
        cur = cur->next;
    }
    return nL;
}

File: CMPS101/pa3/List.java
/*
 * Name: Aaron Steele, atsteele
 * Class: CMPS101
 * Project: Pa3
 */

public class List 
{
    // Private inner classes
    private class Node
    {
        // Fields
        private Node prev, next;
        private Object data;
        
        // Constructors
        // Blank constructor
        public Node()
        {
            prev = null;
            next = null;
            data = null; // eh
        }
        // Node in the middle of the list
        public Node(Object data, Node prev, Node next)
        {
            this.data = data;
            this.prev = prev;
            this.next = next;
        }
        
        // Access functions
        public Object data()
        {
            return data;
        }
        public Node next()
        {
            return next;
        }
        public Node prev()
        {
            return prev;
        }
        
        // Manipulation procedures
        public void setData(Object data)
        {
            this.data = data;
        }
        public void setNext(Node next)
        {
            this.next = next;
        }
        public void setPrev(Node prev)
        {
            this.prev = prev;
        }
        
        // Other methods
        @Override
        public boolean equals(Object x)
        {
        	// Don't need to bother calling .equals if they aren't even the same class
        	if(x instanceof Node)
        	{
        		// Call the object's equals method, since it's an object
        		return ((Node) x).data().equals(this.data);
        	}
        	return false;
        }
        
        @Override
        public String toString()
        {
            return null;
        }
    }
    

    // Fields
    private int index, length;
    private Node front, back, cursor;
    
    
    // Constructors
    public List()
    {
        index = -1;
        length = 0;
        front = null;
        back = null;
        cursor = null;
    }
    
    // Access functions
    /**
     * Length of the list
     * @return
     */
    public int length()
    {
        return length;
    }
    /**
     * Index of cursor. If not defined, -1.
     * @return
     */
    public int index()
    {
        if(cursor != null)
            return index;
        else
            return -1;
    }
    /**
     * Pre: length > 0
     * @return Node at front of list
     */
    public Object front()
    {
        if(length() <= 0)
            throw new RuntimeException("front(): length <= 0");
        return front.data();
    }
    /**
     * Pre: length > 0
     * @return Node at back of list
     */
    public Object back()
    {
        if(length() <= 0)
            throw new RuntimeException("back(): length <= 0");
        return back.data();
    }
    /**
     * Pre: length > 0
     * @return Node at cursor position
     */
    public Object get()
    {
        if(length() <= 0)
            throw new RuntimeException("get(): length <= 0");
        return cursor.data();
    }
    
    @Override
    public boolean equals(Object x)
    {
        Node cur = front;
        Node LCur = ((List) x).front;
        
        while(cur != null)
        {
            // If L is null before our list is, it's necessarily not equal
            if(LCur == null || !LCur.data().equals(cur.data()))
                return false;
            cur = cur.next();
            LCur = LCur.next();
        }
        return true;
    }
    
    // Manipulation procedures
    /**
     * Reset list to empty
     */
    public void clear()
    {
        front = null;
        back = null;
        cursor = null;
        index = -1;
        length = 0;
    }
    /**
     * If list is not empty, put cursor at front element
     * Else, do nothing
     */
    public void moveFront()
    {
        if(length() > 0)
        {
            cursor = front;
            index = 0;
        }
    }
    /**
     * If list is not empty, put cursor at end element
     * Else, do nothing
     */
    public void moveBack()
    {
        if(length() > 0)
        {
            cursor = back;
            index = length() - 1;
        }
    }
    /**
     * If cursor defined, if not at beginning of list, move cursor back. Else, set cursor to undefined. Else, do nothing
     */
    public void movePrev()
    {
        if(cursor != null)
        {
            if(index() > 0)
            {
                cursor = cursor.prev();
                index--;
            }
            else
            {
                cursor = null;
            }
        }
    }
    /**
     * If cursor defined, if not at end of list, move cursor forward. Else, set cursor to undefined. Else, do nothing
     */
    public void moveNext()
    {
        if(cursor != null)
        {
            if(index() != length() - 1)
            {
                cursor = cursor.next();
                index++;
            }
            else
            {
                cursor = null;
            }
        }
    }
    /**
     * Insert new element.
     * If list is non-empty, put before front element
     * @param data
     */
    public void prepend(Object data)
    {
        Node n = new Node(data, null, front);
        if(length() == 0)
            back = n;
        if(front != null)
            front.prev = n;
        front = n;
        length++;
        
    }
    /**
     * Insert new element.
     * If list is non-empty, put after back element
     * @param data
     */
    public void append(Object data)
    {
        Node n = new Node(data, back, null);
        if(length() == 0)
            front = n;
        if(back != null)
            back.next = n;
        back = n;
        length++;
        
    }
    /**
     * Insert new element before cursor
     * Pre: length>0, index>=0
     * @param data
     */
    public void insertBefore(Object data)
    {
        if(length() <= 0)
            throw new RuntimeException("insertBefore(): length <= 0");
        if(index() < 0)
            throw new RuntimeException("insertBefore(): index < 0");
        
        Node n = new Node(data, cursor.prev(), cursor);
        if(cursor.prev() != null)
            cursor.prev().setNext(n);
        else // If there was nothing before the cursor, cursor was at front, and we need to update front
            front = n;
        cursor.setPrev(n);
        length++;
    }
    /**
     * Insert new element after cursor
     * Pre: length>0, index>=0
     * @param data
     */
    public void insertAfter(Object data)
    {
        if(length() <= 0)
            throw new RuntimeException("insertAfter(): length <= 0");
        if(index() < 0)
            throw new RuntimeException("insertAfter(): index < 0");
        
        Node n = new Node(data, cursor, cursor.next());
        if(cursor.next() != null)
            cursor.next().setPrev(n);
        else // If there was nothing after the cursor, cursor was at back, and we need to update back
            back = n;
        cursor.setNext(n);
        length++;
    }
    /**
     * Delete front element
     * Pre: length>0
     */
    public void deleteFront()
    {
        if(length() <= 0)
            throw new RuntimeException("deleteFront(): length <= 0");
        
        // Just dereference the front node. Garbage collection will take care of the rest
        front = front.next();
        front.setPrev(null);
        if(index() == 0) // If cursor was at front, repoint it to the current front
        {
            cursor = front;
        }
        length--;
        
    }
    /**
     * Delete back element
     * Pre: length>0
     */
    public void deleteBack()
    {
        if(length() <= 0)
            throw new RuntimeException("deleteBack(): length <= 0");
        
        // Dereference back node
        back = back.prev();
        back.setNext(null);
        if(index() == length() - 1) // If cursor was at back, repoint it to back
        {
            cursor = back;
        }
        length--;
        index--; // Also decrease index by 1, as we lost 1 element
        
    }
    /**
     * Delete cursor element, making cursor undefined
     * Pre: length>0, index>=0
     */
    public void delete()
    {
        if(length() <= 0)
            throw new RuntimeException("insertAfter(): length <= 0");
        if(index() < 0)
            throw new RuntimeException("insertAfter(): index < 0");
        
        if(cursor.prev() != null)
        	cursor.prev().setNext(cursor.next());
        if(cursor.next() != null)
        	cursor.next().setPrev(cursor.prev());
        length--;

        if(index() == 0) // Cursor was at front, repoint front to new front
        	front = cursor.next();
        if(index() == length() - 1) // Cursor was at back, repoint back to new back
        	back = cursor.prev();
        
        cursor = null;
        if(index() > 0)
            index--; // Correct index
        
    }
    
    // Other methods
    @Override
    public String toString()
    {
        String ret = "";
        Node cur = front;
        while(cur != null)
        {
            ret += cur.data().toString() + " ";
            cur = cur.next();
        }
        ret.trim();
        
        return ret;
    }
}
